Add parameter option to `Request`

konDeichmann opened the pull request:
### Changes

This PR adds the functionality to add parameters on a token-renew.

In `Request.swift` I added a function (`parameters(_ payload: [String: Any])`) that returns a copy of self.
The returned object will contain all fields from before, but with an updated payload.

The functionality is very similar to `BaseWebAuth.parameters(_:)`

This functionality is also exposed via `CredentialsManager.credentials()`.

### References

Linked to #493 

### Testing

Changes in this PR have been tested in a real-world Product. 
Unit Tests have been written to verify that `Request.parameters(_:)` work as expected.

* [x] This change adds unit test coverage
* [ ] This change has been tested on the latest version of the platform/language or why not

### Checklist

* [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
* [x] All existing and new tests complete without errors
* [x] All active GitHub checks have passed

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
     /// - Note: [Auth0 Refresh Tokens Docs](https://auth0.com/docs/tokens/concepts/refresh-tokens)
     #if WEB_AUTH_PLATFORM
-    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
+    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, withParameters parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```
```suggestion
    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
                     return callback(.touchFailed($0!), nil)
                 }
-                self.retrieveCredentials(withScope: scope, minTTL: minTTL, callback: callback)
+                self.retrieveCredentials(withScope: scope, minTTL: minTTL, withParameters: parameters, callback: callback)
```
```suggestion
                self.retrieveCredentials(withScope: scope, minTTL: minTTL, parameters: parameters, callback: callback)
```

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
     /// - Parameters:
     ///   - scope: scopes to request for the new tokens. By default is nil which will ask for the same ones requested during original Auth.
     ///   - minTTL: minimum time in seconds the access token must remain valid to avoid being renewed.
+    ///   - parameters: additional Parameter to add to a possible Token refresh, parameters will be set via Request.payload.
```
```suggestion
    ///   - parameters: additional parameters to add to a possible Token refresh. The parameters will be set via Request.payload.
```

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
             }
         } else {
-            self.retrieveCredentials(withScope: scope, minTTL: minTTL, callback: callback)
+            self.retrieveCredentials(withScope: scope, minTTL: minTTL, withParameters: parameters, callback: callback)
```
```suggestion
            self.retrieveCredentials(withScope: scope, minTTL: minTTL, parameters: parameters, callback: callback)
```

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
     }
     #else
-    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
+    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, withParameters parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```
```suggestion
    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
+    public func credentials(withScope scope: String? = nil, minTTL: Int = 0, withParameters parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
         guard self.hasValid(minTTL: minTTL) else { return callback(.noCredentials, nil) }
-        self.retrieveCredentials(withScope: scope, minTTL: minTTL, callback: callback)
+        self.retrieveCredentials(withScope: scope, minTTL: minTTL, withParameters: parameters, callback: callback)
```
```suggestion
        self.retrieveCredentials(withScope: scope, minTTL: minTTL, parameters: parameters, callback: callback)
```

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
     }
 
-    private func retrieveCredentials(withScope scope: String?, minTTL: Int, callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
+    private func retrieveCredentials(withScope scope: String?, minTTL: Int, withParameters parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```
```suggestion
    private func retrieveCredentials(withScope scope: String?, minTTL: Int, parameters parameters: [String: Any] = [:], callback: @escaping (CredentialsManagerError?, Credentials?) -> Void) {
```

Widcket commented on Auth0/Request.swift:
```diff
@@ -95,6 +95,17 @@ public struct Request<T, E: Auth0Error>: Requestable {
+
+     - parameter payload: Additional Parameters for the Request, provided map will be added to `paylaod`.
+     */
+    public func parameters(_ payload: [String: Any]) -> Self {
```
```suggestion
    public func parameters(_ parameters: [String: Any]) -> Self {
```

Widcket replied:
```suggestion
    public func parameters(_ parameters: [String: Any]) -> Self {
```

Widcket commented on Auth0/Request.swift:
```diff
@@ -95,6 +95,17 @@ public struct Request<T, E: Auth0Error>: Requestable {
+    /**
+     Modify the Parameters, by creating a copy of self and adding the provided parameters to `payload`.
+
+     - parameter payload: Additional Parameters for the Request, provided map will be added to `paylaod`.
```
```suggestion
     - parameter parameters: Additional parameters for the Request. The provided map will be added to `parameters`.
```

Widcket commented on Auth0/Request.swift:
```diff
@@ -95,6 +95,17 @@ public struct Request<T, E: Auth0Error>: Requestable {
     }
 
+    /**
+     Modify the Parameters, by creating a copy of self and adding the provided parameters to `payload`.
```
```suggestion
     Modify the parameters by creating a copy of self and adding the provided parameters to `parameters`.
```

Widcket commented on Auth0/Request.swift:
```diff
@@ -95,6 +95,17 @@ public struct Request<T, E: Auth0Error>: Requestable {
+     - parameter payload: Additional Parameters for the Request, provided map will be added to `paylaod`.
+     */
+    public func parameters(_ payload: [String: Any]) -> Self {
+        var parameter = self.payload
+        payload.forEach { parameter[$0] = $1 }
+
+        return Request(session: self.session, url: self.url, method: self.method, handle: self.handle, payload: parameter, headers: self.headers, logger: self.logger, telemetry: self.telemetry)
+    }
```
```suggestion
        var mutableParameters = self.parameters
        parameters.forEach { mutableParameters[$0] = $1 }

        return Request(session: self.session, url: self.url, method: self.method, handle: self.handle, parameters: mutableParameters, headers: self.headers, logger: self.logger, telemetry: self.telemetry)
    }
```

Widcket commented on Auth0Tests/AuthenticationSpec.swift:
```diff
@@ -318,7 +318,46 @@ class AuthenticationSpec: QuickSpec {
+
+            let refreshToken = UUID().uuidString.replacingOccurrences(of: "-", with: "")
+
+            it("should contain payload") {
+                let request = auth.renew(withRefreshToken: refreshToken)
+
+                expect(request.payload["refresh_token"] as? String) == refreshToken
+                expect(request.payload["grant_type"] as? String) == "refresh_token"
+                expect(request.payload["client_id"] as? String) == ClientId
+            }
```
```suggestion
            it("should contain parameters") {
                let request = auth.renew(withRefreshToken: refreshToken)

                expect(request.parameters["refresh_token"] as? String) == refreshToken
                expect(request.parameters["grant_type"] as? String) == "refresh_token"
                expect(request.parameters["client_id"] as? String) == ClientId
            }
```

Widcket commented on Auth0Tests/AuthenticationSpec.swift:
```diff
@@ -318,7 +318,46 @@ class AuthenticationSpec: QuickSpec {
+                        "phone": Phone
+                    ])
+
+                expect(request.payload["refresh_token"] as? String) == refreshToken
+                expect(request.payload["grant_type"] as? String) == "refresh_token"
+                expect(request.payload["client_id"] as? String) == "new Client ID"
+                expect(request.payload["phone"] as? String) == Phone
```
```suggestion
                expect(request.parameters["refresh_token"] as? String) == refreshToken
                expect(request.parameters["grant_type"] as? String) == "refresh_token"
                expect(request.parameters["client_id"] as? String) == "new Client ID"
                expect(request.parameters["phone"] as? String) == Phone
```

Widcket commented on Auth0Tests/AuthenticationSpec.swift:
```diff
@@ -318,7 +318,46 @@ class AuthenticationSpec: QuickSpec {
+                expect(baseRequest.session) == modifiedRequest.session
+                expect(baseRequest.url) == modifiedRequest.url
+                expect(baseRequest.method) == modifiedRequest.method
+                expect(baseRequest.payload as? [String: String]) == modifiedRequest.payload as? [String: String]
```
```suggestion
                expect(baseRequest.parameters as? [String: String]) == modifiedRequest.parameters as? [String: String]
```

Widcket reviewed (changes requested):
Thank you for raising this PR. Looks great, just some little details.
Please add some tests in `CredentialsManagerSpec`.

konDeichmann commented:
Hi @Widcket, thanks for your feedback. 

I modified the Code according to your comments and added a Test that refreshes the Token with a custom parameter. The test would fail if that Parameter is not present.

I'm not too familiar with the HTTP Library therefore I don't know if my test could be improved. I'm very open for feedback again 😄

Widcket commented on Auth0/CredentialsManager.swift:
```diff
@@ -164,70 +164,74 @@ public struct CredentialsManager {
     /// - Parameters:
     ///   - scope: scopes to request for the new tokens. By default is nil which will ask for the same ones requested during original Auth.
     ///   - minTTL: minimum time in seconds the access token must remain valid to avoid being renewed.
+    ///   - parameters: additional parameters to add to a possible Token refresh. The parameters will be set via Request.payload.
```
```suggestion
    ///   - parameters: additional parameters to add to a possible token refresh. The parameters will be set via Request.payload.
```

Widcket commented on Auth0Tests/CredentialsManagerSpec.swift:
```diff
@@ -530,6 +530,23 @@ class CredentialsManagerSpec: QuickSpec {
                     }
                 }
 
+                it("Request should contain Parameter on renew") {
```
```suggestion
                it("Request should contain parameter on renew") {
```

Widcket commented on Auth0/Request.swift:
```diff
@@ -95,6 +95,17 @@ public struct Request<T, E: Auth0Error>: Requestable {
+    /**
+     Modify the parameters by creating a copy of self and adding the provided parameters to `payload`.
+
+     - parameter payload: Additional parameters for the Request. The provided map will be added to `payload`.
```
```suggestion
     - parameter payload: Additional parameters for the request. The provided map will be added to `payload`.
```

Widcket reviewed (approved):
Thank you!

Widcket commented:
This is now out in version [1.36.0](https://github.com/auth0/Auth0.swift/releases/tag/1.36.0).
